Exploratory Data Analysis

In [1]:
import pandas as pd

In [2]:
import os 
for x in os.listdir(r"C:\Users\Aditya Mithar\Desktop\KP programming\Python\Data Science"):
    if "Sales" in x:
        print(x)

Sales_April_2019.csv
Sales_August_2019.csv
Sales_December_2019.csv
Sales_February_2019.csv
Sales_January_2019.csv
Sales_July_2019.csv
Sales_June_2019.csv
Sales_March_2019.csv
Sales_May_2019.csv
Sales_November_2019.csv
Sales_October_2019.csv


In [3]:
directory_path = r"C:\Users\Aditya Mithar\Desktop\KP programming\Python\Data Science"
dataset = pd.DataFrame()

for x in os.listdir(directory_path):
    if "Sales" in x:
        file_path = os.path.join(directory_path, x)
        d = pd.read_csv(file_path)
        dataset = pd.concat([dataset, d])

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 175164 entries, 0 to 20378
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          174659 non-null  object
 1   Product           174659 non-null  object
 2   Quantity Ordered  174659 non-null  object
 3   Price Each        174659 non-null  object
 4   Order Date        174659 non-null  object
 5   Purchase Address  174659 non-null  object
dtypes: object(6)
memory usage: 9.4+ MB


In [5]:
dataset.isnull().sum()/dataset.shape[0]*100   #%age of the null values

Order ID            0.288301
Product             0.288301
Quantity Ordered    0.288301
Price Each          0.288301
Order Date          0.288301
Purchase Address    0.288301
dtype: float64

In [6]:
dataset.shape

(175164, 6)

In [7]:
dataset.dropna(axis=0, inplace=True)

In [8]:
dataset.shape

(174659, 6)

In [9]:
dataset.to_csv(r"C:\Users\Aditya Mithar\Desktop\KP programming\Python\Data Science\Output.csv", index=False)

In [10]:
dataset = pd.read_csv("Output.csv")

In [11]:
dataset

Order ID                     Product Quantity Ordered Price Each  \
0        176558        USB-C Charging Cable                2      11.95   
1        176559  Bose SoundSport Headphones                1      99.99   
2        176560                Google Phone                1        600   
3        176560            Wired Headphones                1      11.99   
4        176561            Wired Headphones                1      11.99   
...         ...                         ...              ...        ...   
174654   278792       AA Batteries (4-pack)                1       3.84   
174655   278793            Wired Headphones                1      11.99   
174656   278794       AA Batteries (4-pack)                1       3.84   
174657   278795                      iPhone                1        700   
174658   278796            Wired Headphones                1      11.99   

            Order Date                         Purchase Address  
0       04/19/19 08:46             917 1st St, Dallas, TX 75001  
1       04/07/19 22:30        682 Chestnut St, Boston, MA 02215  
2       04/12/19 14:38     669 Spruce St, Los Angeles, CA 90001  
3       04/12/19 14:38     669 Spruce St, Los Angeles, CA 90001  
4       04/30/19 09:27        333 8th St, Los Angeles, CA 90001  
...                ...                                      ...  
174654  10/12/19 04:32    920 Adams St, San Francisco, CA 94016  
174655  10/28/19 22:00   161 Chestnut St, Los Angeles, CA 90001  
174656  10/09/19 20:58   346 Spruce St, San Francisco, CA 94016  
174657  10/31/19 17:21           291 Hill St, Seattle, WA 98101  
174658  10/13/19 11:20  547 Jefferson St, Los Angeles, CA 90001  

[174659 rows x 6 columns]

In [12]:
dataset.head()

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
1   176559  Bose SoundSport Headphones                1      99.99   
2   176560                Google Phone                1        600   
3   176560            Wired Headphones                1      11.99   
4   176561            Wired Headphones                1      11.99   

       Order Date                      Purchase Address  
0  04/19/19 08:46          917 1st St, Dallas, TX 75001  
1  04/07/19 22:30     682 Chestnut St, Boston, MA 02215  
2  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  
4  04/30/19 09:27     333 8th St, Los Angeles, CA 90001

Question 1: Which month is best for sales?

In [13]:
dataset["Product"].value_counts()

Product
USB-C Charging Cable          20449
Lightning Charging Cable      20329
AAA Batteries (4-pack)        19357
AA Batteries (4-pack)         19260
Wired Headphones              17707
Apple Airpods Headphones      14585
Bose SoundSport Headphones    12499
27in FHD Monitor               7028
iPhone                         6445
27in 4K Gaming Monitor         5849
34in Ultrawide Monitor         5804
Google Phone                   5183
Flatscreen TV                  4509
Macbook Pro Laptop             4436
ThinkPad Laptop                3879
20in Monitor                   3827
Vareebadd Phone                1939
LG Washing Machine              620
LG Dryer                        616
Product                         338
Name: count, dtype: int64

Some rows are just storing the name of the columns. Displayig them

In [14]:
dataset[dataset["Product"]== "Product"]

Order ID  Product  Quantity Ordered  Price Each  Order Date  \
517     Order ID  Product  Quantity Ordered  Price Each  Order Date   
1146    Order ID  Product  Quantity Ordered  Price Each  Order Date   
1152    Order ID  Product  Quantity Ordered  Price Each  Order Date   
2869    Order ID  Product  Quantity Ordered  Price Each  Order Date   
2884    Order ID  Product  Quantity Ordered  Price Each  Order Date   
...          ...      ...               ...         ...         ...   
172315  Order ID  Product  Quantity Ordered  Price Each  Order Date   
172355  Order ID  Product  Quantity Ordered  Price Each  Order Date   
172900  Order ID  Product  Quantity Ordered  Price Each  Order Date   
173170  Order ID  Product  Quantity Ordered  Price Each  Order Date   
174583  Order ID  Product  Quantity Ordered  Price Each  Order Date   

        Purchase Address  
517     Purchase Address  
1146    Purchase Address  
1152    Purchase Address  
2869    Purchase Address  
2884    Purchase Address  
...                  ...  
172315  Purchase Address  
172355  Purchase Address  
172900  Purchase Address  
173170  Purchase Address  
174583  Purchase Address  

[338 rows x 6 columns]

Removing the rows that contain just the name of columns

In [15]:
dataset = dataset[dataset["Product"] != "Product"]

In [16]:
dataset.shape

(174321, 6)

In [17]:
dataset["Product"].value_counts()

Product
USB-C Charging Cable          20449
Lightning Charging Cable      20329
AAA Batteries (4-pack)        19357
AA Batteries (4-pack)         19260
Wired Headphones              17707
Apple Airpods Headphones      14585
Bose SoundSport Headphones    12499
27in FHD Monitor               7028
iPhone                         6445
27in 4K Gaming Monitor         5849
34in Ultrawide Monitor         5804
Google Phone                   5183
Flatscreen TV                  4509
Macbook Pro Laptop             4436
ThinkPad Laptop                3879
20in Monitor                   3827
Vareebadd Phone                1939
LG Washing Machine              620
LG Dryer                        616
Name: count, dtype: int64

In [18]:
dataset.dtypes

Order ID            object
Product             object
Quantity Ordered    object
Price Each          object
Order Date          object
Purchase Address    object
dtype: object

In [19]:
#changing the data types of quantity ordered and price each to numeric
dataset["Quantity Ordered"] = pd.to_numeric(dataset["Quantity Ordered"])
dataset["Price Each"] = pd.to_numeric(dataset["Price Each"])

C:\Users\Aditya Mithar\AppData\Local\Temp\ipykernel_12124\2216202461.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset["Quantity Ordered"] = pd.to_numeric(dataset["Quantity Ordered"])
C:\Users\Aditya Mithar\AppData\Local\Temp\ipykernel_12124\2216202461.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset["Price Each"] = pd.to_numeric(dataset["Price Each"])


In [20]:
dataset.dtypes

Order ID             object
Product              object
Quantity Ordered      int64
Price Each          float64
Order Date           object
Purchase Address     object
dtype: object

In [21]:
#declaring a new column to store (quantity ordered)*(Price Each)
dataset["Sales"] = dataset["Price Each"] * dataset["Quantity Ordered"]

C:\Users\Aditya Mithar\AppData\Local\Temp\ipykernel_12124\3580103907.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset["Sales"] = dataset["Price Each"] * dataset["Quantity Ordered"]


In [22]:
dataset

Order ID                     Product  Quantity Ordered  Price Each  \
0        176558        USB-C Charging Cable                 2       11.95   
1        176559  Bose SoundSport Headphones                 1       99.99   
2        176560                Google Phone                 1      600.00   
3        176560            Wired Headphones                 1       11.99   
4        176561            Wired Headphones                 1       11.99   
...         ...                         ...               ...         ...   
174654   278792       AA Batteries (4-pack)                 1        3.84   
174655   278793            Wired Headphones                 1       11.99   
174656   278794       AA Batteries (4-pack)                 1        3.84   
174657   278795                      iPhone                 1      700.00   
174658   278796            Wired Headphones                 1       11.99   

            Order Date                         Purchase Address   Sales  
0       04/19/19 08:46             917 1st St, Dallas, TX 75001   23.90  
1       04/07/19 22:30        682 Chestnut St, Boston, MA 02215   99.99  
2       04/12/19 14:38     669 Spruce St, Los Angeles, CA 90001  600.00  
3       04/12/19 14:38     669 Spruce St, Los Angeles, CA 90001   11.99  
4       04/30/19 09:27        333 8th St, Los Angeles, CA 90001   11.99  
...                ...                                      ...     ...  
174654  10/12/19 04:32    920 Adams St, San Francisco, CA 94016    3.84  
174655  10/28/19 22:00   161 Chestnut St, Los Angeles, CA 90001   11.99  
174656  10/09/19 20:58   346 Spruce St, San Francisco, CA 94016    3.84  
174657  10/31/19 17:21           291 Hill St, Seattle, WA 98101  700.00  
174658  10/13/19 11:20  547 Jefferson St, Los Angeles, CA 90001   11.99  

[174321 rows x 7 columns]

In [23]:
dataset.head()

Order ID                     Product  Quantity Ordered  Price Each  \
0   176558        USB-C Charging Cable                 2       11.95   
1   176559  Bose SoundSport Headphones                 1       99.99   
2   176560                Google Phone                 1      600.00   
3   176560            Wired Headphones                 1       11.99   
4   176561            Wired Headphones                 1       11.99   

       Order Date                      Purchase Address   Sales  
0  04/19/19 08:46          917 1st St, Dallas, TX 75001   23.90  
1  04/07/19 22:30     682 Chestnut St, Boston, MA 02215   99.99  
2  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  600.00  
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001   11.99  
4  04/30/19 09:27     333 8th St, Los Angeles, CA 90001   11.99

In [24]:
#changing the dtype of order data to srting
dataset["Order Date"].str

In [25]:
dataset["Order Date"].dtype

dtype('O')

In [26]:
#slicing the order date string to just get the month
dataset["Order Date"].str[:2]

0         04
1         04
2         04
3         04
4         04
          ..
174654    10
174655    10
174656    10
174657    10
174658    10
Name: Order Date, Length: 174321, dtype: object

In [27]:
#storing the sliced date string to a new varible
dataset["Month"] = dataset["Order Date"].str[:2]

C:\Users\Aditya Mithar\AppData\Local\Temp\ipykernel_12124\633071283.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset["Month"] = dataset["Order Date"].str[:2]


In [28]:
dataset.head()

Order ID                     Product  Quantity Ordered  Price Each  \
0   176558        USB-C Charging Cable                 2       11.95   
1   176559  Bose SoundSport Headphones                 1       99.99   
2   176560                Google Phone                 1      600.00   
3   176560            Wired Headphones                 1       11.99   
4   176561            Wired Headphones                 1       11.99   

       Order Date                      Purchase Address   Sales Month  
0  04/19/19 08:46          917 1st St, Dallas, TX 75001   23.90    04  
1  04/07/19 22:30     682 Chestnut St, Boston, MA 02215   99.99    04  
2  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  600.00    04  
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001   11.99    04  
4  04/30/19 09:27     333 8th St, Los Angeles, CA 90001   11.99    04

In [29]:
dataset["Month"].value_counts()
#can also be written as:
#dataset.Month.value_counts()

Month
12    24984
10    20262
04    18279
11    17573
05    16566
03    15153
07    14293
06    13554
02    11975
08    11961
01     9709
09       12
Name: count, dtype: int64

In [30]:
#summing up the month wise sales

dataset.groupby("Month")["Sales"].sum()

Month
01    1822256.73
02    2202022.42
03    2807100.38
04    3390670.24
05    3152606.75
06    2577802.26
07    2647775.76
08    2244467.88
09       2096.74
10    3733373.57
11    3199603.20
12    4613443.34
Name: Sales, dtype: float64

In [31]:
#sorting the sales in descending order

dataset.groupby("Month")["Sales"].sum().sort_values(ascending=False)

Month
12    4613443.34
10    3733373.57
04    3390670.24
11    3199603.20
05    3152606.75
03    2807100.38
07    2647775.76
06    2577802.26
08    2244467.88
02    2202022.42
01    1822256.73
09       2096.74
Name: Sales, dtype: float64

Thus, December has maximum sales.

Question 2: Which city sold most product?

In [32]:
dataset["Purchase Address"]

0                    917 1st St, Dallas, TX 75001
1               682 Chestnut St, Boston, MA 02215
2            669 Spruce St, Los Angeles, CA 90001
3            669 Spruce St, Los Angeles, CA 90001
4               333 8th St, Los Angeles, CA 90001
                           ...                   
174654      920 Adams St, San Francisco, CA 94016
174655     161 Chestnut St, Los Angeles, CA 90001
174656     346 Spruce St, San Francisco, CA 94016
174657             291 Hill St, Seattle, WA 98101
174658    547 Jefferson St, Los Angeles, CA 90001
Name: Purchase Address, Length: 174321, dtype: object

In [33]:
#Every city name is separeted using comma.
#Separating the purchase address on comma.
dataset["Purchase Address"].str.split(",")

0                    [917 1st St,  Dallas,  TX 75001]
1               [682 Chestnut St,  Boston,  MA 02215]
2            [669 Spruce St,  Los Angeles,  CA 90001]
3            [669 Spruce St,  Los Angeles,  CA 90001]
4               [333 8th St,  Los Angeles,  CA 90001]
                             ...                     
174654      [920 Adams St,  San Francisco,  CA 94016]
174655     [161 Chestnut St,  Los Angeles,  CA 90001]
174656     [346 Spruce St,  San Francisco,  CA 94016]
174657             [291 Hill St,  Seattle,  WA 98101]
174658    [547 Jefferson St,  Los Angeles,  CA 90001]
Name: Purchase Address, Length: 174321, dtype: object

The above data is a list which contains street on its 0th index and city name on 1st.

In [34]:
dataset["Purchase Address"].str.split(",").str[1]

0                 Dallas
1                 Boston
2            Los Angeles
3            Los Angeles
4            Los Angeles
               ...      
174654     San Francisco
174655       Los Angeles
174656     San Francisco
174657           Seattle
174658       Los Angeles
Name: Purchase Address, Length: 174321, dtype: object

In [35]:
#storing this data in a variable named city
dataset["City"] = dataset["Purchase Address"].str.split(",").str[1]

C:\Users\Aditya Mithar\AppData\Local\Temp\ipykernel_12124\1862289858.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset["City"] = dataset["Purchase Address"].str.split(",").str[1]


In [36]:
dataset.head()

Order ID                     Product  Quantity Ordered  Price Each  \
0   176558        USB-C Charging Cable                 2       11.95   
1   176559  Bose SoundSport Headphones                 1       99.99   
2   176560                Google Phone                 1      600.00   
3   176560            Wired Headphones                 1       11.99   
4   176561            Wired Headphones                 1       11.99   

       Order Date                      Purchase Address   Sales Month  \
0  04/19/19 08:46          917 1st St, Dallas, TX 75001   23.90    04   
1  04/07/19 22:30     682 Chestnut St, Boston, MA 02215   99.99    04   
2  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  600.00    04   
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001   11.99    04   
4  04/30/19 09:27     333 8th St, Los Angeles, CA 90001   11.99    04   

           City  
0        Dallas  
1        Boston  
2   Los Angeles  
3   Los Angeles  
4   Los Angeles

In [37]:
dataset["City"].unique()

array([' Dallas', ' Boston', ' Los Angeles', ' San Francisco', ' Seattle',
       ' Atlanta', ' New York City', ' Portland', ' Austin'], dtype=object)

In [38]:
dataset.groupby("City")["Sales"].sum().sort_values(ascending=False)

City
 San Francisco    7797674.26
 Los Angeles      5098352.38
 New York City    4363026.66
 Boston           3414273.28
 Atlanta          2624219.69
 Dallas           2603759.55
 Seattle          2591401.16
 Portland         2187541.13
 Austin           1712971.16
Name: Sales, dtype: float64

Thus, San Francisco sold the most.

Question 3: Which time is best for selling.

In [39]:
dataset["Order Date"]

0         04/19/19 08:46
1         04/07/19 22:30
2         04/12/19 14:38
3         04/12/19 14:38
4         04/30/19 09:27
               ...      
174654    10/12/19 04:32
174655    10/28/19 22:00
174656    10/09/19 20:58
174657    10/31/19 17:21
174658    10/13/19 11:20
Name: Order Date, Length: 174321, dtype: object

In [40]:
#obtaing the time of each sale
dataset["Order Date"].str[-5:]

0         08:46
1         22:30
2         14:38
3         14:38
4         09:27
          ...  
174654    04:32
174655    22:00
174656    20:58
174657    17:21
174658    11:20
Name: Order Date, Length: 174321, dtype: object

In [41]:
#obtaining which hour sold the most
dataset["hour"] = dataset["Order Date"].str[-5:-3]

C:\Users\Aditya Mithar\AppData\Local\Temp\ipykernel_12124\1487118778.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset["hour"] = dataset["Order Date"].str[-5:-3]


In [42]:
dataset.head()

Order ID                     Product  Quantity Ordered  Price Each  \
0   176558        USB-C Charging Cable                 2       11.95   
1   176559  Bose SoundSport Headphones                 1       99.99   
2   176560                Google Phone                 1      600.00   
3   176560            Wired Headphones                 1       11.99   
4   176561            Wired Headphones                 1       11.99   

       Order Date                      Purchase Address   Sales Month  \
0  04/19/19 08:46          917 1st St, Dallas, TX 75001   23.90    04   
1  04/07/19 22:30     682 Chestnut St, Boston, MA 02215   99.99    04   
2  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  600.00    04   
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001   11.99    04   
4  04/30/19 09:27     333 8th St, Los Angeles, CA 90001   11.99    04   

           City hour  
0        Dallas   08  
1        Boston   22  
2   Los Angeles   14  
3   Los Angeles   14  
4   Los Angeles   09

In [43]:
dataset["hour"].value_counts()

hour
19    12112
12    11812
11    11650
18    11535
20    11506
13    11365
10    10314
14    10252
21    10169
17    10164
16     9762
15     9567
22     8236
09     8185
08     5881
23     5855
07     3794
00     3650
06     2338
01     2186
05     1232
02     1165
04      815
03      776
Name: count, dtype: int64

In [44]:
dataset.groupby("hour")["Sales"].sum().sort_values(ascending=False)

hour
19    2262362.07
12    2197187.82
11    2163313.93
20    2150274.05
18    2088234.69
13    2033395.80
17    1981725.49
14    1939195.94
21    1913954.36
15    1832220.34
10    1827461.94
16    1781786.15
09    1531630.20
22    1510735.08
08    1125015.65
23    1102927.71
07     711726.42
00     662083.31
01     432834.21
06     420604.97
05     217845.95
02     213413.28
04     152388.36
03     140901.55
Name: Sales, dtype: float64

Thus, the sales are best in the evening time from 5-6 pm onwards, being  best at 7pm. While 11am to 1pm is also considerablly good time of sales.

Question 4: What product sold the most?

In [45]:
#which product is ordered maximum no. of time?
dataset.groupby("Product")["Quantity Ordered"].sum().sort_values(ascending=False)

Product
AAA Batteries (4-pack)        29090
AA Batteries (4-pack)         25862
USB-C Charging Cable          22381
Lightning Charging Cable      21784
Wired Headphones              19255
Apple Airpods Headphones      14693
Bose SoundSport Headphones    12627
27in FHD Monitor               7068
iPhone                         6451
27in 4K Gaming Monitor         5862
34in Ultrawide Monitor         5821
Google Phone                   5190
Flatscreen TV                  4525
Macbook Pro Laptop             4440
ThinkPad Laptop                3881
20in Monitor                   3852
Vareebadd Phone                1942
LG Washing Machine              620
LG Dryer                        616
Name: Quantity Ordered, dtype: int64

Thus, AAA battries sold the most.

Question 5: What product sold the most? Why do you think it sold the most?

In [46]:
dataset.groupby("Product")["Price Each"].mean().sort_values()

Product
AAA Batteries (4-pack)           2.99
AA Batteries (4-pack)            3.84
USB-C Charging Cable            11.95
Wired Headphones                11.99
Lightning Charging Cable        14.95
Bose SoundSport Headphones      99.99
20in Monitor                   109.99
27in FHD Monitor               149.99
Apple Airpods Headphones       150.00
Flatscreen TV                  300.00
34in Ultrawide Monitor         379.99
27in 4K Gaming Monitor         389.99
Vareebadd Phone                400.00
Google Phone                   600.00
LG Washing Machine             600.00
LG Dryer                       600.00
iPhone                         700.00
ThinkPad Laptop                999.99
Macbook Pro Laptop            1700.00
Name: Price Each, dtype: float64

Hence, AAA battries sold the most since its the chepest product.

Question 6: What products are most often sold together?

In [47]:
#People who buy multiple products together will have the same order id

dataset["Order ID"].duplicated

<bound method Series.duplicated of 0         176558
1         176559
2         176560
3         176560
4         176561
           ...  
174654    278792
174655    278793
174656    278794
174657    278795
174658    278796
Name: Order ID, Length: 174321, dtype: object>

In [48]:
dataset["Order ID"].duplicated().sum()

#These many order ids are repeated in our dataset

7091

In [49]:
data = dataset[dataset["Order ID"].duplicated(keep=False)]

In [50]:
data

Order ID                     Product  Quantity Ordered  Price Each  \
2        176560                Google Phone                 1      600.00   
3        176560            Wired Headphones                 1       11.99   
17       176574                Google Phone                 1      600.00   
18       176574        USB-C Charging Cable                 1       11.95   
29       176585  Bose SoundSport Headphones                 1       99.99   
...         ...                         ...               ...         ...   
174565   278706        USB-C Charging Cable                 1       11.95   
174569   278710        USB-C Charging Cable                 1       11.95   
174570   278710            27in FHD Monitor                 1      149.99   
174578   278718        USB-C Charging Cable                 1       11.95   
174579   278718  Bose SoundSport Headphones                 1       99.99   

            Order Date                      Purchase Address   Sales Month  \
2       04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  600.00    04   
3       04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001   11.99    04   
17      04/03/19 19:42     20 Hill St, Los Angeles, CA 90001  600.00    04   
18      04/03/19 19:42     20 Hill St, Los Angeles, CA 90001   11.95    04   
29      04/07/19 11:31     823 Highland St, Boston, MA 02215   99.99    04   
...                ...                                   ...     ...   ...   
174565  10/07/19 18:18  942 11th St, San Francisco, CA 94016   11.95    10   
174569  10/13/19 19:05   183 5th St, San Francisco, CA 94016   11.95    10   
174570  10/13/19 19:05   183 5th St, San Francisco, CA 94016  149.99    10   
174578  10/13/19 12:45     479 Lincoln St, Atlanta, GA 30301   11.95    10   
174579  10/13/19 12:45     479 Lincoln St, Atlanta, GA 30301   99.99    10   

                  City hour  
2          Los Angeles   14  
3          Los Angeles   14  
17         Los Angeles   19  
18         Los Angeles   19  
29              Boston   11  
...                ...  ...  
174565   San Francisco   18  
174569   San Francisco   19  
174570   San Francisco   19  
174578         Atlanta   12  
174579         Atlanta   12  

[13823 rows x 10 columns]

In [51]:
data["Grouped"] = data.groupby("Order ID")["Product"].transform(lambda x: ", ".join(x))

C:\Users\Aditya Mithar\AppData\Local\Temp\ipykernel_12124\1248883651.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data["Grouped"] = data.groupby("Order ID")["Product"].transform(lambda x: ", ".join(x))


In [52]:
data.head()

Order ID                     Product  Quantity Ordered  Price Each  \
2    176560                Google Phone                 1      600.00   
3    176560            Wired Headphones                 1       11.99   
17   176574                Google Phone                 1      600.00   
18   176574        USB-C Charging Cable                 1       11.95   
29   176585  Bose SoundSport Headphones                 1       99.99   

        Order Date                      Purchase Address   Sales Month  \
2   04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  600.00    04   
3   04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001   11.99    04   
17  04/03/19 19:42     20 Hill St, Los Angeles, CA 90001  600.00    04   
18  04/03/19 19:42     20 Hill St, Los Angeles, CA 90001   11.95    04   
29  04/07/19 11:31     823 Highland St, Boston, MA 02215   99.99    04   

            City hour                                            Grouped  
2    Los Angeles   14                     Google Phone, Wired Headphones  
3    Los Angeles   14                     Google Phone, Wired Headphones  
17   Los Angeles   19                 Google Phone, USB-C Charging Cable  
18   Los Angeles   19                 Google Phone, USB-C Charging Cable  
29        Boston   11  Bose SoundSport Headphones, Bose SoundSport He...

In [53]:
data = data[["Order ID", "Grouped"]].drop_duplicates()

In [54]:
data.head()

Order ID                                            Grouped
2     176560                     Google Phone, Wired Headphones
17    176574                 Google Phone, USB-C Charging Cable
29    176585  Bose SoundSport Headphones, Bose SoundSport He...
31    176586               AAA Batteries (4-pack), Google Phone
118   176672     Lightning Charging Cable, USB-C Charging Cable

In [55]:
data["Grouped"].value_counts().sort_values(ascending=False)[:10]

Grouped
iPhone, Lightning Charging Cable                        834
Google Phone, USB-C Charging Cable                      798
iPhone, Wired Headphones                                343
Vareebadd Phone, USB-C Charging Cable                   292
Google Phone, Wired Headphones                          288
iPhone, Apple Airpods Headphones                        274
Google Phone, Bose SoundSport Headphones                150
Vareebadd Phone, Wired Headphones                       100
Google Phone, USB-C Charging Cable, Wired Headphones     75
Vareebadd Phone, Bose SoundSport Headphones              55
Name: count, dtype: int64

Hence, as we can see iphone and lightning cable or google phone and usb-c charging cables are the most sold together products.

Question 7: How much probability for the next people will order USB-C charging cable?

In [56]:
#formula for probablity: frequency/total frequency

def probability(frequency, total_frequency):
    return frequency/total_frequency

In [57]:
total = dataset["Product"].value_counts().sum()
total

174321

In [58]:
#getting the count of how much usb-c cable is sold
usb_c = dataset[dataset["Product"] == "USB-C Charging Cable"]["Quantity Ordered"].value_counts().sum()
usb_c

20449

In [59]:
probability(usb_c, total)

0.11730657809443498

Hence, the required prob. is 0.1173

Question 8: How much probablity for next people  will buy other products?

In [60]:
for x in dataset["Product"].unique():
    other_products = dataset[dataset["Product"] == x]["Quantity Ordered"].value_counts().sum()
    print(x, ": ", probability(other_products, total))

USB-C Charging Cable :  0.11730657809443498
Bose SoundSport Headphones :  0.07170105724496761
Google Phone :  0.02973250497645149
Wired Headphones :  0.10157697580899605
Macbook Pro Laptop :  0.02544730697965248
Lightning Charging Cable :  0.11661819287406566
27in 4K Gaming Monitor :  0.033553042949501205
AA Batteries (4-pack) :  0.11048582786927565
Apple Airpods Headphones :  0.0836674869923876
AAA Batteries (4-pack) :  0.11104227258907418
iPhone :  0.03697202287733549
Flatscreen TV :  0.025866074655377148
27in FHD Monitor :  0.04031642773962976
20in Monitor :  0.021953751986278188
LG Dryer :  0.003533710797895836
ThinkPad Laptop :  0.022252052248438226
Vareebadd Phone :  0.011123157852467574
LG Washing Machine :  0.0035566569719081465
34in Ultrawide Monitor :  0.033294898491862714


The above result represents the required probability.